0000-navigator-view-api.md

```diff
@@ -0,0 +1,89 @@
+* Start Date: 2018-01-29
+* RFC PR: (leave this empty)
+* React Navigation Issue: (leave this empty)
+
+# Summary
+
+Introduce a simpler API for createNavigator and navigation views. By providing "sceneDescriptors" to the navigation view, it does not need to compute options or manually create the child navigation prop.
+
+Down the line, this will enable us to deprecate `addNavigationHelpers`, which is an avoidable burden for redux apps and when implementing custom navigators. Eventually, each navigator may provide a different set of helpers.
+
+# Basic example
+
+Using a new `createNavigator` API and `screenDescriptors` prop, this is simplest possible custom navigation view:
+
+```js
+const CustomNavigationView = ({navigation, screenDescriptors}) => {
+  const {routes, index} = navigation;
+  const descriptor = screenDescriptors[routes[index].key];
+  const ScreenComponent = descriptor.getComponent();
+  return (
```

Comment:
I find the way this screen element is constructed could be easier to understand if it were a function that rendered the element. Ex:
```js
const screen = nullthrows(screens[routes[index].key]);
return screen.renderScreenElement({ navigation });


// where we have defined earlier:
screen.renderScreenElement(props) {
  return <HomeScreen someOption={whatever} {...props} />;
}
```

This has a couple of nice properties:
1. Some of the RFC's proposed descriptor properties like `options` are coupled to the screen element. Within `renderScreenElement()` the options and the element are now more cohesive.
2. There is less back-and-forth between React Navigation and the application code.
3. Props aren't bundled into an `options` (or `params` or other synonym for `stuff`).

Reply:
I think we still need *something* like `screenDescriptor.options`, because the navigation view needs to access several properties from it, such as "headerTitle" in the case of StackView.

Reply:
Also, one thing to note, is that it can be convenient to have access to ScreenComponent directly. In the case of drawer navigator's current implementation of this RFC, it is necessary to access the `ScreenComponent.router` (Although I think we should change this to make the DrawerNavigator easier to understand).

Reply:
It seems like you could get all of that with an API like this:
```
var screen = {
  renderScreenElement(props) {
    return <HomeScreen {...props} />;
  },
  getNavigatorOptions() {
    return { headerTitle, etc };
  },
  getRouter() {
    return router;
  },
}
```

For screen components that need the navigator options or router, you can easily modify the code so that HomeScreen has access to those variables. But this is pushed into user space and is relying on plain JS / programming logic more than React Navigation's framework API.

Reply:
By "screen" I assume you mean "sceneDescriptor"? Earlier today Brent and I decided that "descriptor" is probably the best term to use for now.

I'm a bit confused who would be defining the actual <HomeScreen /> element. Are you suggesting a top-level API like this?

```
const App = StackNavigator({
  Home: {
    renderScreenElement: (props) => {
      const HomeScreen = require('HomeScreen');
      return <HomeScreen {...props} custom={options} />;
    },
    getNavigatorOptions() {
      return { headerTitle, etc };
    },
    getRouter() {
      const HomeScreen = require('HomeScreen');
      return HomeScreen.router;
    },
  },
  Profile: {
    ...
```

Reply:
Yeah, same thing as `sceneDescriptor`. And yeah I suppose this would be a user space API change too, making it a better candidate for later. The key thing is that information shared across different properties of the descriptor is easier to read because it's all in one place and is just JavaScript, hiding less logic inside of React Navigation or its docs. Ex with the current API I'd say this is directionally better:

```js
StackNavigator({
  Home: {
    getComponent() {
      return ({ navigation, options }) => <HomeScreen navigation={navigation} options={options} />;
    }
    options: { ... },
    navigation: { ... },
  }
});
```

compared to this:
```js
StackNavigator({
  Home: {
    getComponent() {
      return HomeScreen;
    }
    options: { ... },
    navigation: { ... },
  }
});
```

The reason is that the first snippet is a pathway towards removing some logic (namely the flow of some props) that's hiding within React Navigation:

```js
let options = { ... };
let navigation = { ... };
StackNavigator({
  Home: {
    getComponent() {
      return () => <HomeScreen navigation={navigation} options={options} />;
    }
    options,
    navigation,
  }
});
```

Now you can see where the `navigation` and `options` props are coming from if you understand JS or any other lexically scoped language. The exact syntax can be massaged a bit (I find it's a bit less desirable grouping of concerns to define `options` for all your screens up top) but that's the gist.

Reply:
I can see some value in letting users render screens themselves, because it will help them understand what is going on. But some of the logic locked up in react-nav provides a lot of value, for example the auto-creation of navigation props. The logic in `let navigation = { ... };` would push a lot of boilerplate to the users.

This would be a good followup RFC, and we will also need to figure out how to iterate on these APIs without causing too many breaking changes for end users.